File: src/com/nvv/cache/java/AbstractCache.java
package com.nvv.cache.java;

import java.util.HashMap;
import java.util.Map;

public abstract class AbstractCache<K, V> {

    private static final int MAX_ITEMS_IN_CACHE = 3;

    protected Map<K, CacheElement<K, V>> mTable = new HashMap<K, CacheElement<K, V>>();

    protected boolean isFull() {
        return itemsCount() >= MAX_ITEMS_IN_CACHE;
    }

    protected CacheElement<K, V> createCacheElement(K key, V value) {
        CacheElement<K, V> cacheElement = new CacheElement<K, V>();
        cacheElement.setKey(key);
        cacheElement.setValue(value);
        return cacheElement;
    }

    protected boolean updateCacheElement(CacheElement<K, V> element, K key, V value) {
        if (element != null) {
            element.setValue(value);
            element.setKey(key);

            return true;
        }

        return false;
    }

    public abstract void put(K key, V value);

    public abstract V get(K key);

    public abstract void print();

    protected abstract int itemsCount();
}

File: src/com/nvv/cache/kotlin/AbstractCache.kt
package com.nvv.cache.kotlin

import java.util.HashMap
import java.util.Map

public abstract class AbstractCache<K, V> {

    private val MAX_ITEMS_IN_CACHE : Int = 3

    protected val mTable : Map<K, CacheElement<K, V>> = HashMap<K, CacheElement<K, V>>()

    protected fun isFull() : Boolean {
        return itemsCount() >= MAX_ITEMS_IN_CACHE;
    }

    protected fun createCacheElement(key : K, value : V) : CacheElement<K, V> {
        val cacheElement = CacheElement<K, V>()
        cacheElement.key = key;
        cacheElement.value = value;
        return cacheElement;
    }

    protected fun updateCacheElement(element : CacheElement<K, V>? , key : K , value : V) : Boolean {
        if (element != null) {
            element.key = key;
            element.value = value;
            return true;
        }

        return false;
    }

    public abstract fun put(key : K, value :V)

    public abstract fun get(key : K) : V?

    public abstract fun print()

    protected abstract fun itemsCount() : Int
}